### PR TITLE
Experiments With GPT-2

### PR DESCRIPTION
**Pull request by spacemanidol**

### CODE REVIEW

File: research/GPT/helpers.py [hunk 1]
```diff
@@ -0,0 +1,883 @@
...
+    :return: a list of NamedLayerParam tuples whose full parameter names in the given
+        module match one of the given regex patterns or parameter names
+    """
+    named_layers_and_params = []
+    found_param_names = []
+    import pdb
+    
+    for layer_name, layer in module.named_modules():
+        for param_name, param in layer.named_parameters():
+            if "." in param_name and ".bias" not in param_name and ".weight" not in param_name:  # skip parameters of nested layers
```
Comment (spacemanidol) [inline research/GPT/helpers.py:778]: @bfineran

File: research/GPT/helpers.py [hunk 2]
```diff
@@ -0,0 +1,880 @@
...
+
+import numpy
+import torch
+from torch import Tensor
+from torch.nn import Linear, Module, Parameter
+from torch.nn.modules.conv import _ConvNd
+from torch.optim.optimizer import Optimizer
+from torch.utils.data import DataLoader
+
+from transformers.modeling_utils import Conv1D
```
Comment (spacemanidol) [inline research/GPT/helpers.py:34]: temporary work around because gpt2 has custom convs

File: research/GPT/helpers.py [hunk 3]
```diff
@@ -0,0 +1,880 @@
...
+def get_prunable_layers(module: Module) -> List[Tuple[str, Module]]:
+    """
+    :param module: the module to get the prunable layers from
+    :return: a list containing the names and modules of the prunable layers
+        (Linear, ConvNd)
+    """
+    layers = []
+
+    for name, mod in module.named_modules():
+        if isinstance(mod, Linear) or isinstance(mod, _ConvNd) or isinstance(mod, Conv1D):
```
Comment (spacemanidol) [inline research/GPT/helpers.py:739]: work around

File: research/GPT/recipes/90sparse.yaml [hunk 4]
```diff
@@ -0,0 +1,152 @@
+version: 1.1.0
+
+modifiers:
+    - !EpochRangeModifier
+        end_epoch: 10
+        start_epoch: 0.0
+
+        - !GMPruningModifier
```
Comment (bfineran) [inline research/GPT/recipes/90sparse.yaml:8]: is there a reason that there are so many `GMPruningModifier`s here? it looks like they generally all have the same hyperparameters

File: research/GPT/requirements.txt [hunk 5]
```diff
@@ -0,0 +1,4 @@
+torch
+sparseml
+transformers
+numpy
```
Comment (bfineran) [inline research/GPT/requirements.txt:4]: nit: `sparseml[torch]` will make sure a compatible version of pytorch is installed, and numpy should come with them
```suggestion
sparseml[torch]
transformers
```

File: research/GPT/src/run_language_modeling.py [hunk 6]
```diff
@@ -0,0 +1,546 @@
...
+        export_model(model, eval_dataloader, data_args.onnx_export_path)
+
+
+def _mp_fn(index):
+    # For xla_spawn (TPUs)
+    main()
+
+
+if __name__ == "__main__":
+    main()
```
Comment (bfineran) [inline research/GPT/src/run_language_modeling.py:546]: nit: no newline at end of file

File: research/GPT/src/sparseml_utils.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    Module exporter class for Sequence Classification
+    """
+    @classmethod
+    def get_output_names(self, out: Any):
+        if not isinstance(out, CausalLMOutput):
+            raise ValueError("Expected CausalLMOutput, got {type(out)}")
+        expected = ["logits"]
+        if numpy.any([name for name in expected if name not in out]):
+            raise ValueError("Expected output names not found in model output")
+        return expected
```
Comment (bfineran) [inline research/GPT/src/sparseml_utils.py:63]: nit: no newline at end of file

**Review by bfineran — COMMENTED** [review]
great work @spacemanidol, excited to see this in action

**Pull request closed without merging**